### PR TITLE
sql: rationalize the handling of comments and improve performance

### LINKED ISSUE #102851: 75% increase in Django test suite run time between CockroachDB 22.2 and 23.1
**Describe the problem**

The full Django test suite as run by TeamCity for https://github.com/cockroachdb/django-cockroachdb/ takes around 40 minutes for CockroachDB 22.2.x. 

The same run as of CockroachDB 23.1 rc2 takes around 70 minutes, down from around 90 minutes at 23.1 beta 3, before the fix for https://github.com/cockroachdb/cockroach/issues/100871.

You can see the various builds at https://github.com/cockroachdb/django-cockroachdb/pull/279.

(I couldn't identify the cause of the regression during the 23.1 development cycle as there were months when the Django test suite wouldn't run without error due to https://github.com/cockroachdb/cockroach/issues/94337.)

I generated a slow query log with threshold 1000ms for CockraochDB 22.2.8 and CockroachDB nightly (v23.1.0-alpha.8-dev-3f8cb919339a7f19bd3d20c86488d6a46af9fb7f @ 2023/05/05 23:26:11) (the nightly build gives a similar total run time as 23.1 rc2.) 

22.2.8: 423 queries with total time of 8.7 minutes
Nightly: 460 queries with total time of 10.1 minutes

This result suggests to me that there may be some other cause besides simply slow queries (though it's possible there are a lot of queries with sub-1-second run time that are accounting for the difference). I'm happy to continue investigating with a pointer in the right direction, or perhaps it would be more efficient for the CockroachDB developers to take it from here.

Jira issue: CRDB-27689

### PR DESCRIPTION
**Pull request by rafiss**
rebase of https://github.com/cockroachdb/cockroach/pull/95431
fixes https://github.com/cockroachdb/cockroach/issues/102851

Prior to this patch, we had a weird translation going in
`kv_catalog_comments`, which was causing the tree of objects to be
traversed 6 times (one per comment type), and then the rows to be
mapped one by one again in `pg_description` / `pg_shdescription`.

This made no sense - it should be possible to push a predicate on
`pg_description` down to a virtual index on `kv_catalog_comments`, and
this requires the column types to align and no computation to happen
in-between.

This commit simplifies all this by having `kv_catalog_comments`
responsible for preparing the data in the format expected by the `pg_`
tables.

The number of roundtrips needed for an ORM query that uses
pg_description dropped to 4, from 134.

```
goos: linux
goarch: amd64
cpu: Intel(R) Xeon(R) CPU @ 2.80GHz
                                             │ /tmp/tmp.0yUMg6OauX/bench.6abe1851d83 │ /tmp/tmp.0yUMg6OauX/bench.89a494eb3010b945ae5a2be7a057e50802b231c0 │
                                             │                sec/op                 │                   sec/op                    vs base                │
ORMQueries/introspection_description_join-24                           10687.1m ± 1%                                  144.1m ± 8%  -98.65% (p=0.000 n=10)

                                             │ /tmp/tmp.0yUMg6OauX/bench.6abe1851d83 │ /tmp/tmp.0yUMg6OauX/bench.89a494eb3010b945ae5a2be7a057e50802b231c0 │
                                             │              roundtrips               │                 roundtrips                  vs base                │
ORMQueries/introspection_description_join-24                            134.000 ± 0%                                   4.000 ± 0%  -97.01% (p=0.000 n=10)

                                             │ /tmp/tmp.0yUMg6OauX/bench.6abe1851d83 │ /tmp/tmp.0yUMg6OauX/bench.89a494eb3010b945ae5a2be7a057e50802b231c0 │
                                             │                 B/op                  │                    B/op                     vs base                │
ORMQueries/introspection_description_join-24                          3846.69Mi ± 0%                                 38.98Mi ± 6%  -98.99% (p=0.000 n=10)

                                             │ /tmp/tmp.0yUMg6OauX/bench.6abe1851d83 │ /tmp/tmp.0yUMg6OauX/bench.89a494eb3010b945ae5a2be7a057e50802b231c0 │
                                             │               allocs/op               │                 allocs/op                   vs base                │
ORMQueries/introspection_description_join-24                           45566.9k ± 0%                                  433.3k ± 6%  -99.05% (p=0.000 n=10)
```

Release note: None

### CODE REVIEW

File: pkg/sql/logictest/testdata/logic_test/information_schema [hunk 1]
```diff
@@ -2949,20 +2952,20 @@ SELECT table_catalog, table_schema, table_name, column_name, column_comment
...
-table_catalog  table_schema  table_name  column_name    column_comment
-test           public        t1          id             'identification'
-test           public        t1          name           'character'
-test           public        t1          class          ·
-test           public        t1          rowid          ·
+table_catalog  table_schema  table_name  column_name  column_comment
+test           public        t1          id           identification
+test           public        t1          name         character
+test           public        t1          class        NULL
+test           public        t1          rowid        NULL
```
Comment (postamar) [inline pkg/sql/logictest/testdata/logic_test/information_schema:2959]: Do we want NULLs here, or should those be replaced by empty strings? I honestly can't remember exactly. Feel free to disregard if this doesn't matter.

**Review by postamar — COMMENTED** [review]
Very nice!

File: pkg/sql/crdb_internal.go [hunk 2]
```diff
@@ -5443,41 +5475,56 @@ CREATE TABLE crdb_internal.kv_catalog_comments (
...
+				db, err := p.Descriptors().ByIDWithLeased(p.txn).Get().Database(ctx, schema.GetParentID())
+				if err != nil {
+					return err
+				}
+				c, err := catalog.MustFindConstraintByID(tableDesc, descpb.ConstraintID(key.SubID))
+				if err != nil {
+					return err
+				}
+				classOid = tree.NewDOid(catconstants.PgCatalogConstraintTableID)
+				objOid = getOIDFromConstraint(c, db.GetID(), schema.GetID(), tableDesc)
```
[the remark this answers is not in the document]
Comment (rafiss) [reply pkg/sql/crdb_internal.go:5518]: i had just copied the logic from `pkg/sql/pg_catalog.go` so i can't take credit for the nice idea :)

the issue you're describing is pre-existing, but i can fix it

**Review by cucaroach — COMMENTED** [review]
Seems fine from the bits I understood.   Nice speedup!

> Reviewed 7 of 21 files at r1, 10 of 15 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/104763#-:-NXyyiNy2w1zpkaLnLM3:b-amblnl)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @fqazi, @postamar, and @rafiss)

___
*[`-- commits` line 42 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104763#-NXytEvJ-pmy2T1jyZRz):*
Not sure what happened here but looks like some whitespace got eaten or something.

**Review by rafiss — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/104763)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach, @fqazi, and @postamar)

___
*[`-- commits` line 42 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104763#-NXytEvJ-pmy2T1jyZRz):*
<details><summary><i>Previously, cucaroach (Tommy Reilly) wrote…</i></summary><blockquote>

Not sure what happened here but looks like some whitespace got eaten or something.
</blockquote></details>



hm, it seems fine to me. you might need to make your browser window wider to view the table easily.

**Review by rafiss — COMMENTED** [review]
@postamar this is ready for another look

**Review by postamar — APPROVED** [review]

**Comment by rafiss** [comment]
tftr!

bors r+

**Pull request merged**